File: nextfrontend/src/api/cv.js
/**
 * CV API client.
 *
 * CV data is stored as a JSONB blob in the database.
 * Image uploads (profile pic inside CV) use presigned MinIO URLs.
 */
import apiClient from './client';
import { uploadFileViaPresigned } from './storage';

export const getCVDataApi = async (language) => {
  const params = language ? { language } : {};
  const { data } = await apiClient.get('/cv/', { params });
  return data;
};

export const updateCVDataApi = async (cvData, language) => {
  const params = language ? { language } : {};
  const { data } = await apiClient.put('/cv/', cvData, { params });
  return data;
};

/**
 * AI-assisted CV import: upload a CV/resume file (PDF, image or text) and
 * have the configured Gemini model extract it into the CV JSON structure.
 *
 * The result is returned for review (not saved automatically) - call
 * `updateCVDataApi` afterwards to persist it.
 *
 * @param {File}   file     - Browser File object (PDF/PNG/JPEG/WEBP/TXT)
 * @param {string} mode     - 'replace' (fresh data) or 'merge' (extend existing data)
 * @param {string} language - CV language to merge into / save as
 * @returns {Promise<object>} Generated CV data
 */
export const importCVApi = async (file, mode, language) => {
  const formData = new FormData();
  formData.append('file', file);
  formData.append('mode', mode);

  const params = language ? { language } : {};
  const { data } = await apiClient.post('/cv/import', formData, {
    params,
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 120000,
  });
  return data;
};

/**
 * Upload a CV-related image via presigned URL.
 *
 * @param {File}        file       - Browser File object
 * @param {string}      [filename] - Desired filename (e.g. "profile.webp")
 * @param {number|null} [projectId] - Optional project reference
 * @returns {Promise<{ object_name: string }>}
 */
export const uploadCVImageApi = async (file, filename, projectId = null) => {
  // Rename the file if a custom filename was provided
  const renamedFile = filename
    ? new File([file], filename, { type: file.type })
    : file;
  return uploadFileViaPresigned(renamedFile, 'cv', projectId);
};

/** Alias for backwards compatibility */
export const uploadImageApi = uploadCVImageApi;

File: backend/src/api/routers/cv.py
"""CV (Curriculum Vitae) endpoints."""

from fastapi import APIRouter, Depends, File, Form, UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.dependencies import get_current_admin_user, get_db
from ...db.models.user import User
from ...services import cv as cv_service
from ..schemas.cv import CVData

router = APIRouter(prefix="/cv", tags=["cv"])


@router.get("/", response_model=CVData)
async def read_cv(
    language: str = "en",
    db: AsyncSession = Depends(get_db),
):
    """Public endpoint - returns the CV data without authentication."""
    return await cv_service.get_cv_data(db, language=language)


@router.put("/", response_model=CVData)
async def update_cv(
    data: CVData,
    language: str = "en",
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    return await cv_service.update_cv_data(
        db, data=data, owner_id=current_user.id, language=language
    )


@router.post("/import", response_model=CVData)
async def import_cv(
    file: UploadFile = File(...),
    mode: str = Form(...),
    language: str = "en",
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    AI-assisted CV import (admin only).

    Uploads a CV/resume file (PDF, image, or text) and uses the configured
    Gemini model to extract its content into the CV JSON structure.

    ``mode="replace"`` generates fresh CV data from the file alone.
    ``mode="merge"`` merges the file's content into the existing CV data
    for ``language``.

    The result is returned for review - it is NOT saved automatically.
    Use ``PUT /cv/`` to persist the reviewed data.
    """
    return await cv_service.import_cv_from_file(
        db, file=file, mode=mode, language=language
    )
